account for uncertainty using deadband

wholmgren opened the pull request:
- [x] Closes #358 
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Initial attempt at accounting for uncertainty using a deadband. Example code in details below. @alorenzo175 let me know what you think of this approach before I clean it up. I'm trying to leave some flexibility for the dashboard to say "use this deadband for all or use preset uncertainty for all". In the long run it might be nice to let the dashboard set the deadband for each, but I think that would require more refactoring in `calculator` than is in scope here.

<details>

```python
latitude = 32.2
longitude = -110.9
elevation = 700

site = datamodel.Site(
    name='Tucson, AZ',
    latitude=latitude,
    longitude=longitude,
    elevation=elevation,
    timezone='America/Phoenix'
)

name = 'OASIS DA GHI'
variable = 'ghi'

fx1 = datamodel.Forecast(
    name=name + ' 15',
    site=site,
    variable=variable,
    interval_label='ending',
    interval_value_type='mean',
    interval_length=pd.Timedelta('1h'),
    issue_time_of_day=datetime.time(hour=0, minute=45),
    run_length=pd.Timedelta('1h'),
    lead_time_to_start=pd.Timedelta('15min'),
    forecast_id='',
    extra_parameters='',
)

fx2 = datamodel.Forecast(
    name=name + ' 75',
    site=site,
    variable=variable,
    interval_label='ending',
    interval_value_type='mean',
    interval_length=pd.Timedelta('1h'),
    issue_time_of_day=datetime.time(hour=0, minute=45),
    run_length=pd.Timedelta('1h'),
    lead_time_to_start=pd.Timedelta('75min'),
    forecast_id='',
    extra_parameters='',
)

obs = datamodel.Observation(
    name=name,
    site=site,
    variable=variable,
    interval_label='ending',
    interval_value_type='mean',
    interval_length=pd.Timedelta('1h'),
    uncertainty=1
)

fxobs1 = datamodel.ForecastObservation(
    forecast=fx1,
    observation=obs
)

fxobs2 = datamodel.ForecastObservation(
    forecast=fx2,
    observation=obs
)

index = pd.date_range('20200401T1900', freq='1h', periods=3, tz='UTC')

obs_data = pd.DataFrame({'value': [1, 1, 1], 'quality_flag': 2}, index=index)
fx1_data = pd.Series([1, 2, 3], index=index)
fx2_data = pd.Series([0, 1, 2], index=index)

data = {
    fx1: fx1_data,
    fx2: fx2_data,
    obs: obs_data
}

filters = [datamodel.QualityFlagFilter()]

procs = preprocessing.process_forecast_observations([fxobs1, fxobs2], filters, data, 'UTC')

categories = ('total', )
metrics = ('mae', 'mbe', 'ksi')

metrics_result = calculator.calculate_metrics(procs, categories, metrics)

print('default\n', metrics_result)

metrics_result = calculator.calculate_metrics(procs, categories, metrics, deadband=1)

print('\nset deadband\n', metrics_result)

metrics_result = calculator.calculate_metrics(procs, categories, metrics, deadband='observation_uncertainty')

print('\ninfer deadband\n', metrics_result)

default
 [MetricResult(name='OASIS DA GHI 15', forecast_id='', values=(MetricValue(category='total', metric='mae', index='0', value=1.0), MetricValue(category='total', metric='mbe', index='0', value=1.0), MetricValue(category='total', metric='ksi', index='0', value=1.0)), observation_id='', aggregate_id=None), MetricResult(name='OASIS DA GHI 75', forecast_id='', values=(MetricValue(category='total', metric='mae', index='0', value=0.6666666666666666), MetricValue(category='total', metric='mbe', index='0', value=0.0), MetricValue(category='total', metric='ksi', index='0', value=0.6666666666666665)), observation_id='', aggregate_id=None)]

set deadband
 [MetricResult(name='OASIS DA GHI 15', forecast_id='', values=(MetricValue(category='total', metric='mae', index='0', value=0.6666666666666666), MetricValue(category='total', metric='mbe', index='0', value=0.6666666666666666), MetricValue(category='total', metric='ksi', index='0', value=1.0)), observation_id='', aggregate_id=None), MetricResult(name='OASIS DA GHI 75', forecast_id='', values=(MetricValue(category='total', metric='mae', index='0', value=0.0), MetricValue(category='total', metric='mbe', index='0', value=0.0), MetricValue(category='total', metric='ksi', index='0', value=0.6666666666666665)), observation_id='', aggregate_id=None)]

infer deadband
 [MetricResult(name='OASIS DA GHI 15', forecast_id='', values=(MetricValue(category='total', metric='mae', index='0', value=0.6666666666666666), MetricValue(category='total', metric='mbe', index='0', value=0.6666666666666666), MetricValue(category='total', metric='ksi', index='0', value=1.0)), observation_id='', aggregate_id=None), MetricResult(name='OASIS DA GHI 75', forecast_id='', values=(MetricValue(category='total', metric='mae', index='0', value=0.0), MetricValue(category='total', metric='mbe', index='0', value=0.0), MetricValue(category='total', metric='ksi', index='0', value=0.6666666666666665)), observation_id='', aggregate_id=None)]
```

</details>

alorenzo175 commented:
Adding the deadband kwarg to `calculate_metrics` is fine by me for now.

For the metrics functions, I had something in mind more like this
```python
# could be replaced with a simple deadband error fnc then use partials to set deadband
def generate_deadband_error_fnc(deadband):
    def _deadband_error(obs, fx):
        error = fx - obs
        mask = deadband_mask(obs, fx, deadband)
        error = np.where(mask, 0, error)
        return error
    return _deadband_error

def _error(obs, fx):
    return fx - obs

def mean_absolute(obs, fx, error_fnc=_error):
    error = error_fcn(obs, fx)
    return np.mean(np.abs(error))
    
```
 and then use the deadband like

```python
mae = mean_absolute(obs, fx, error_fnc=generate_deadband_error_fnc(deadband))
```
If we ever decide a different error function would be useful, it would be a drop in replacement

alorenzo175 commented:
actually using `functools.partial` is better than my deadband function since it is probably easier to inspect the function and deadband value

wholmgren commented:
Yeah, I remember discussing passing a function and thinking it was reasonable at the time. But I don't think an implementation based on `error = fx - obs` can be generalized to support other useful approaches to handling uncertainty, so I don't see the advantage of the more complicated code. Did you have something in mind?

alorenzo175 commented:
> Yeah, I remember discussing passing a function and thinking it was reasonable at the time. But I don't think an implementation based on `error = fx - obs` can be generalized to support other useful approaches to handling uncertainty, so I don't see the advantage of the more complicated code. Did you have something in mind?

No, but I conceptually prefer that the error metric functions apply some transformations to an error series

alorenzo175 commented:
I guess that is the same thing your ``_error_deadband`` does, but using a switch in that function vs just passing a function that only acts on obs and fx feels bad

wholmgren commented:
I was going to push back but I thought the metrics are "mean absolute **error**" and "root mean squared **error**" etc. so maybe passing a function that defines **error** is fair. Proposed pattern:

```python

def error(obs, fx):
    return fx - obs


def error_deadband(obs, fx, deadband):
    error = fx - obs
    # only apply deadband if deadband != 0 and is not None
    if deadband:
        mask = deadband_mask(obs, fx, deadband)
        error = np.where(mask, 0, error)
    return error


def mean_absolute(obs, fx, error_fnc=error):
    """Mean absolute error (MAE).

    .. math:: \\text{MAE} = 1/n \\sum_{i=1}^n |\\text{fx}_i - \\text{obs}_i|

    Parameters
    ----------
    obs : (n,) array-like
        Observed values.
    fx : (n,) array-like
        Forecasted values.
    error_fnc : function
        First argument is obs, second argument is fx.

    Returns
    -------
    mae : float
        The MAE of the forecast.

    Examples
    --------
    >>> obs = np.array([1, 2, 3, 4])
    >>> fx = np.array([2, 2.04, 1, 3.96])
    >>> mean_absolute(obs, fx)
    0.52
    >>> error_fnc = partial(error_deadband, deadband=0.05)
    >>> mean_absolute(obs, fx, error_fnc=error_fnc)
    0.5
    """
    error = error_fnc(obs, fx)
    return np.mean(np.abs(error))

```

lboeman commented:
>  I'm trying to leave some flexibility for the dashboard to say "use this deadband for all or use preset uncertainty for all". In the long run it might be nice to let the dashboard set the deadband for each, but I think that would require more refactoring in `calculator` than is in scope here.

Since this is being implemented on the dashboard, we'll need to handle it in the API. I had assumed a string field called `uncertainty` would be added to the ForecastObservation objects. Is that the idea here? Just want to make sure that the API gets updated appropriately and we can update the datamodel so that things don't break if those fields start coming back from the API.

wholmgren commented:
@lboeman yeah, I came to that realization as I was writing the whatsnew note https://github.com/SolarArbiter/solarforecastarbiter-core/pull/378/commits/6158e0f61f930561dd3b808c4ee7806e0ead37ca#diff-75c68040af308e8075fa06d26de1a09bR76-R78

I think the following changes are needed to this PR:

* remove `deadband` kwarg from `calculate_metrics`
* remove `deadband` kwarg from `calculate_deterministic_metrics`
* add `uncertainty` attribute to `ForecastObservation`. `[None, float, str]`
* add `uncertainty` attribute to `ProcessedForecastObservation`. `[None, float, str]`
* `process_forecast_observations` propagates the attribute from `ForecastObservation` to `ProcessedForecastObservation`. Presumably no change needed for now.
* `calculate_deterministic_metrics` looks up `ProcessedForecastObservation.uncertainty` and passes to metrics functions.

What do you guys think @lboeman @alorenzo175?

lboeman commented:
Sounds good to me. You may need to add some special handling for uncertainty, since you'll get a string from the api for floats.

wholmgren commented:
hrm, should `ForecastObservation.__post_init__` convert str `'observation_uncertainty'` into a float or is that str handling and look up handled later by the metrics code?

alorenzo175 commented:
tough one. __post_init__is easier to see how it's used vs digging into the metrics code, but then you don't know if the value was set to that via a float input or 'observation_uncertainty'. I would lean towards post init

wholmgren commented:
Yeah, maybe we can address that ambiguity in the report text. Also in this case `ProcessedForecastObservation.uncertainty` is probably just `[float, None]`

wholmgren commented:
Another issue: what do we do for `ForecastAggregate`? I suggest `[float, None]`. https://github.com/SolarArbiter/solarforecastarbiter_dashboard/pull/220 would need to disallow or hide the 'use observation uncertainty' option when an aggregate is selected.

lboeman commented:
> Another issue: what do we do for `ForecastAggregate`? I suggest `[float, None]`. [SolarArbiter/solarforecastarbiter_dashboard#220](https://github.com/SolarArbiter/solarforecastarbiter_dashboard/pull/220) would need to disallow or hide the 'use observation uncertainty' option when an aggregate is selected. I'd consider adding constituent observation uncertainties together in quadrature in a future issue.

Cool, I'll update to disallow uncertainty selection for aggregates for the time being.

wholmgren commented on solarforecastarbiter/io/api.py:
```diff
@@ -775,13 +775,15 @@ def process_report_dict(self, rep_dict):
             fx = self.get_forecast(o['forecast'])
             norm = o.get('normalization')
+            unc = o.get('uncertainty')
             if 'observation' in o:
                 obs = self.get_observation(o['observation'])
                 pair = datamodel.ForecastObservation(
-                    fx, obs, normalization=norm)
+                    fx, obs, normalization=norm, uncertainty=unc)
             elif 'aggregate' in o:
                 agg = self.get_aggregate(o['aggregate'])
-                pair = datamodel.ForecastAggregate(fx, agg, normalization=norm)
+                pair = datamodel.ForecastAggregate(
+                    fx, agg, normalization=norm, uncertainty=unc)
```
@lboeman since we're not using `from_dict`, I need to handle None/float/str types here, right? I think this needs to be done for `normalization` too.

lboeman replied:
Yeah, You should only get a string or None back from the API.  So you'd have to handle the cast to float here or in `__set_uncertainty__` I feel like the latter might make more sense since we're already checking if the field is a string and maybe we should check for the [0, 100] bounds too?

wholmgren replied:
Thanks. For normalization it turns out I'm already doing this in `__set_normalization__` 🤦‍♂  my brain doesn't work anymore.

wholmgren commented:
Current status:

- [ ] implemented deadband in essential metrics functions, including tests. needs final(?) review
- [ ] implemented `uncertainty` in `datamodel` and `api`, including tests. needs initial review
- [ ] added support to `metrics.preprocessing` and `metrics.calculator`, but needs tests
- [ ] add some indication of when uncertainty is used or not in reports. maybe add to the `obsfx_table.html` under "Aligned Pairs"? Also add normalization? Or make a new table with this data?

@alorenzo175 can you give the first two items a review and your thoughts on the 4th?

alorenzo175 commented:
>  add some indication of when uncertainty is used or not in reports. maybe add to the `obsfx_table.html` under "Aligned Pairs"? Also add normalization? Or make a new table with this data?

I'm in favour of a new table. I think people should understand the uncertainty and how it is used (deadband error for now) and a separate table is probably better for getting that across.

alorenzo175 commented on solarforecastarbiter/tests/test_datamodel.py:
```diff
@@ -752,3 +752,10 @@ def test_ForecastObservation_normalization_dc(single_forecast_dc_observation):
     assert np.isnan(single_forecast_wind_speed_observation.normalization)
+
+
+def test_ForecastObservation_uncertainty_invalid(
```
tests for valid making sure properly set?

alorenzo175 commented on solarforecastarbiter/datamodel.py:
```diff
@@ -960,6 +971,26 @@ def __set_aggregate_normalization__(self):
+            # not something we can coerce to a float
+            pass
+        else:
+            object.__setattr__(self, 'uncertainty', unc)
```
should this verify between 0 and 100?

wholmgren replied:
We don't verify the values of most numeric attributes in datamodel.

alorenzo175 commented on solarforecastarbiter/metrics/deterministic.py:
```diff
@@ -5,7 +5,66 @@
+    fx : (n,) array_like
+        Forecasted values.
+    deadband : float
+        Fractional tolerance
```
```suggestion
        Fractional tolerance relative to the observed values.
```

wholmgren commented:
Added/improved tests. Added new table. Hopefully getting close now.

This screenshot shows the table for an example report that has a few types of normalization and uncertainty parameters. A few things to note:

1. Metrics values are generally consistent with uncertainty (different sites, so not immediately obvious for small uncertainty).
2. Apparently the reference site has 0 observation uncertainty - probably will want to fix that.
2. GHI pairs have a normalization of `nan` by default, and the manually specified normalization allows for non-NaN metrics. The report json spec follows.
2. Table might look better with different alignment (probably center) but didn't want to put in the effort here since it might depend on things like number of decimals and window width.
2. Suggestions for better text are welcome.

<img width="860" alt="Screen Shot 2020-04-08 at 7 30 55 PM" src="https://user-images.githubusercontent.com/4383303/78851757-7d8ff900-79cf-11ea-8e0f-e5000241b750.png">

<details>
```
{
    "report_parameters": {
        "name": "test",
        "start": "2019-08-25T00:00:00Z",
        "end": "2019-09-03T00:00:00Z",
        "metrics": [
            "mae",
            "nmae",
            "rmse",
            "nrmse",
            "mbe",
            "nmbe",
            "cpi"
        ],
        "categories": [
            "total",
            "year",
            "month",
            "hour",
            "date"
        ],
        "object_pairs": [
            {
                "observation": "9f657636-7e49-11e9-b77f-0a580a8003e9",
                "forecast": "31ae8548-9a07-11e9-b2f4-0a580a8003d5",
                "uncertainty": "10",
                "normalization": "1000"
            },
            {
                "observation": "93d0c01a-7e49-11e9-a9fd-0a580a8003e9",
                "forecast": "84f06a0c-9c22-11e9-9e5c-0a580a8003d4",
                "uncertainty": "observation_uncertainty"
            },
            {
                "observation": "9f657636-7e49-11e9-b77f-0a580a8003e9",
                "forecast": "77be10e8-9c22-11e9-a974-0a580a8003d5",
                "uncertainty": "100"
            }
      ]
    }
  }
```

wholmgren commented:
Might want to add reference forecasts to this metadata table/description too. I think that's a separate PR though since it will require a lot of work to refactor and especially test `metrics.calculator`.
